### PR TITLE
Enable mergeOptions for composition revisions

### LINKED ISSUE #2690: MergeOption: appendSlice is overwriting instead of appending
### What happened?

Hi all, i guess the mergeOption: appendSlice is broken.
Seems like the appendSlice is not respected and the tagset is overwritten.
Is this a known issue?
Thanks

### How can we reproduce it?
I have the following composition with patches

```yaml
  resources:
    - base: 
        apiVersion: s3.aws.crossplane.io/v1beta1
        kind: Bucket
        metadata:
          name: METADATA_UID
        spec:
          deletionPolicy: Delete
          providerConfigRef:
            name: aws-provider-config
          forProvider:
            acl: private
            locationConstraint: REGION
            versioningConfiguration:
              status: Enabled
            serverSideEncryptionConfiguration:
              rules:
                - applyServerSideEncryptionByDefault:
                    sseAlgorithm: AES256
            publicAccessBlockConfiguration:
              blockPublicAcls: true
              blockPublicPolicy: true
              ignorePublicAcls: true
              restrictPublicBuckets: true
            tagging:
              tagSet: 
                - key: namespace
                  value: MY_NAMESPACE
          writeConnectionSecretToRef:
            name: METADATA_NAME-MR-s3bucket-managed-resource
            namespace: crossplane-system
      patches:
        - type: FromCompositeFieldPath
          fromFieldPath: spec.parameters.tags
          toFieldPath: spec.forProvider.tagging.tagSet
          policy:
            mergeOptions:
              appendSlice: true
              keepMapValues: true
```

My claim has the following:
```yaml
    tags:
      - key: another-tag
        value: with-value
```

So the following would be expected result:
```yaml
            tagging:
              tagSet: 
                - key: namespace
                  value: MY_NAMESPACE
                - key: another-tag
                  value: with-value
```

But i get the following:
```yaml
    tagging:
      tagSet:
      - key: another-tag
        value: with-value
```

### What environment did it happen in?
Crossplane version: 

Crossplane version: 1.4.1

Cloud provider or hardware configuration - aws-provider v0.20.0-rc
Kubernetes version (use kubectl version) - v1.21.2-eks-0389ca3
Kubernetes distribution (e.g. Tectonic, GKE, OpenShift) - EKS
OS (e.g. from /etc/os-release) - Amazon Linux

### PR DESCRIPTION
**Pull request by eloo-abi**
Signed-off-by: Joseph Weigl <joseph.weigl@audi.de>

### Description of your changes

This PR will add the mergeOptions to composition revision mechanism

Fixes #2690 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

I have used the S3 template for provider-aws mentioned in #2690

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/apiextensions/v1alpha1/revision_types.go [hunk 1]
```diff
@@ -256,6 +256,8 @@ type Patch struct {
 // A FromFieldPathPolicy determines how to patch from a field path.
 type FromFieldPathPolicy string
 
+type MergeOptionsPolicy xpv1.MergeOptions
```
Comment (negz) [inline apis/apiextensions/v1alpha1/revision_types.go:259]: I'd prefer to define our own types here. You can see some discussion around this topic at https://github.com/crossplane/crossplane/pull/3315.

**Review by negz — COMMENTED** [review]
Thanks @eloo-abi! Apart from one comment this looks great.

File: apis/apiextensions/v1alpha1/revision_types.go [hunk 2]
```diff
@@ -272,6 +274,7 @@ type PatchPolicy struct {
 	// +optional
 	// +immutable
 	FromFieldPath *FromFieldPathPolicy `json:"fromFieldPath,omitempty"`
+	MergeOptions  *MergeOptionsPolicy  `json:"mergeOptions,omitempty"`
```
Comment (negz) [inline apis/apiextensions/v1alpha1/revision_types.go:277]:
```suggestion
	MergeOptions  *xpv1.MergeOptions  `json:"mergeOptions,omitempty"`
```

Would this work? I believe it does the same thing, but more closely mirrors the implementation in the `Composition` type.

File: apis/apiextensions/v1alpha1/revision_types.go (hunk 1 above)
Comment (negz) [reply apis/apiextensions/v1alpha1/revision_types.go:259]: > But im not sure what you mean directly.

Sorry, disregard my comment - I had misunderstood what you were doing.


> Are there any plans to merge https://github.com/crossplane/crossplane/pull/3315 in time or will this be more a mid-term target?

I'm not completely on-board with the approach taken by #3315 of importing v1 types into a v1alpha package. I believe the types should be decoupled across versions, even if it results in a bunch of duplication and copying logic.

File: apis/apiextensions/v1alpha1/revision_types.go (hunk 2 above)
Comment (eloo-abi) [reply apis/apiextensions/v1alpha1/revision_types.go]: @negz i have remove the type casting 
is this what you was thinking about?

**Review by negz — APPROVED** [review]

**Comment by negz** [comment]
Thanks @eloo-abi!

**Review by negz — APPROVED** [review]

**Comment by negz** [comment]
@muvaf Would you mind using your code owner powers to merge this? It won't let me because I'm not a code owner. 🙃

**Review by hasheddan — APPROVED** [review]
ping @muvaf :)

**Review by muvaf — APPROVED** [review]

**Pull request merged**